### PR TITLE
Draft of Version 1.0

### PR DESCRIPTION
**Pull request by erezsh**
Changes so far

- Removed deprecated code
- Removed most of the code for Python 2 compatibility.

### CODE REVIEW

**Comment by erezsh** [comment]
Tasks remaining:

- Merge .pyi files into the code, and remove them.
- Update the interface to be more self-consistent (ideas?)
- Improve the code with Python3 features, where possible (optional/low priority)
- Anything else?

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -6,7 +6,7 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        python-version: [2.7, 3.5, 3.6, 3.7, 3.8, 3.9.0-rc - 3.9, pypy2, pypy3]
+        python-version: [3.5, 3.6, 3.7, 3.8, 3.9.0-rc - 3.9, pypy3]
```
Comment (MegaIng) [inline .github/workflows/tests.yml:9]: Didn't we say we start support with 3.6? To get more of the new features (`typing`, f-strings)

File: lark/exceptions.py [hunk 2]
```diff
@@ -193,7 +191,7 @@ def __init__(self, token, expected, considered_rules=None, state=None, interacti
         # TODO considered_rules and expected can be figured out using state
         self.line = getattr(token, 'line', '?')
         self.column = getattr(token, 'column', '?')
-        self.pos_in_stream = getattr(token, 'pos_in_stream', None)
+        self.pos_in_stream = getattr(token, 'start_pos', None)
```
Comment (MegaIng) [inline lark/exceptions.py:194]: We can just rename the attribute.

File: lark/lark.py [hunk 3]
```diff
@@ -1,14 +1,12 @@
 from __future__ import absolute_import
 
-
-from lark.exceptions import ConfigurationError, assert_config
-
+from abc import ABC, abstractmethod
 import sys, os, pickle, hashlib
 from io import open
```
Comment (MegaIng) [inline lark/lark.py:5]: Not needed in python3 AFAIK.

File: lark/tools/standalone.py [hunk 4]
```diff
@@ -185,9 +180,6 @@ def main():
         parser.print_help(sys.stderr)
         sys.exit(1)
     ns = parser.parse_args()
-    if ns.old_start is not None:
-        warn('The syntax `python -m lark.tools.standalone <grammar-file> <start>` is deprecated. Use the -s option')
```
Comment (MegaIng) [inline lark/tools/standalone.py:189]: This should als be reflected in the parser itself.

File: lark/utils.py [hunk 5]
```diff
@@ -222,34 +201,12 @@ def dedup_list(l):
...
-try:
-    compare = cmp
-except NameError:
-    def compare(a, b):
-        if a == b:
-            return 0
-        elif a > b:
-            return 1
-        return -1
+def compare(a, b):
```
Comment (MegaIng) [inline lark/utils.py:204]: Is there a good reason why we are using this function anywhere?

**Review by MegaIng — CHANGES_REQUESTED** [review]
A few initial comments. I will take another look or too and maybe even just implement everything mentioned here in my own commits.

File: .github/workflows/tests.yml (hunk 1 above)
Comment (erezsh) [reply .github/workflows/tests.yml:9]: I wasn't sure, so I left it for now.

I think 3.6+ is probably the right idea, but I wonder if there's anyone still stuck on 3.5.

File: lark/exceptions.py (hunk 2 above)
Comment (erezsh) [reply lark/exceptions.py:194]: You mean, in the exceptions?

I'm less concerned about that (and anyway, `start_pos` wouldn't be a good name in this case). But I'm open to suggestions.

File: lark/tools/standalone.py (hunk 4 above)
Comment (erezsh) [reply lark/tools/standalone.py:189]: How do you mean?

File: lark/utils.py (hunk 5 above)
Comment (erezsh) [reply lark/utils.py:204]: Couldn't find any. Might be a relic that we forgot to remove.

File: lark/tools/standalone.py (hunk 4 above)
Comment (MegaIng) [reply lark/tools/standalone.py:189]: A few lines above this, there is a call to `parser.add_arguments('old_start')`

File: lark/lark.py (hunk 3 above)
Comment (erezsh) [reply lark/lark.py:5]: Why not?

Comment (erezsh) [reply lark/lark.py:5]: Nevermind, right, it's just an alias.

Comment (MegaIng) [reply lark/lark.py:5]: https://docs.python.org/3/library/io.html#io.open

File: .github/workflows/tests.yml (hunk 1 above)
Comment (ThatXliner) [reply .github/workflows/tests.yml:9]: https://endoflife.date/python. I'd recommend starting with 3.6 (it was my personal choice for most of my projects, no regrets)

Comment (MegaIng) [reply .github/workflows/tests.yml:9]: (Also, according to the [stats](https://pypistats.org/packages/lark-parser), python3.5 is used far less than 2.7)

**Comment by ThatXliner** [comment]
About the type-merging, you can possibly use https://github.com/ilevkivskyi/com2ann to merge type comments to type annotations

**Comment by MegaIng** [comment]
@ThatXliner We are not using Type comments. We have separate stub files.

**Comment by ThatXliner** [comment]
> @ThatXliner We are not using Type comments. We have separate stub files.

Oh right. My bad.

I found https://github.com/ambv/retype but it wasn't what I remembered. Hope it helps, though

File: lark/exceptions.py (hunk 2 above)
Comment (MegaIng) [reply lark/exceptions.py:194]: Fair enough. Keep `pos_in_stream`

**Review by MegaIng — COMMENTED** [review]
(I think setup.py also needs to be updated for `py36`)

**Comment by michael-k** [comment]
> Anything else?

Remove

```ini
[bdist_wheel]
universal = 1
```

from setup.cfg.  It's intended for wheels that support both Python 2 and 3.

**Comment by chanicpanic** [comment]
Here is a list of compatibility/deprecated code I have come across in case you want to remove some of it.

- [visitors.py:81](https://github.com/lark-parser/lark/blob/be3be1815ead4064e76ce70af70d87088c77c047/lark/visitors.py#L81
): `__visit_tokens__` backwards compatibility
- [lark.py:306](https://github.com/lark-parser/lark/blob/be3be1815ead4064e76ce70af70d87088c77c047/lark/lark.py#L306
): Python 2.7 compatibility
- [lark.py:371](https://github.com/lark-parser/lark/blob/be3be1815ead4064e76ce70af70d87088c77c047/lark/lark.py#L371
): Old `resolve__antiscore_sum` check
- [lark.py:403](https://github.com/lark-parser/lark/blob/be3be1815ead4064e76ce70af70d87088c77c047/lark/lark.py#L403
): Possible deprecation of lexer callbacks
- [exceptions.py:92](https://github.com/lark-parser/lark/blob/be3be1815ead4064e76ce70af70d87088c77c047/lark/exceptions.py#L92
): Default value of `use_accepts` is for backwards compatibility
- [exceptions.py:233](https://github.com/lark-parser/lark/blob/be3be1815ead4064e76ce70af70d87088c77c047/lark/exceptions.py#L233
): Possible deprecation of `UnexpectedToken.expected`
- [load_grammar.py:819](https://github.com/lark-parser/lark/blob/be3be1815ead4064e76ce70af70d87088c77c047/lark/load_grammar.py#L819
): Python 2.7 code?
- [load_grammar.py:1070](https://github.com/lark-parser/lark/blob/be3be1815ead4064e76ce70af70d87088c77c047/lark/load_grammar.py#L1070
): Python 2.7 compatibility
- [utils.py:127](https://github.com/lark-parser/lark/blob/be3be1815ead4064e76ce70af70d87088c77c047/lark/utils.py#L127
): Issue with `wraps` in 2.7

**Comment by erezsh** [comment]
Fixes some of the items @chanicpanic mentioned in this PR - https://github.com/lark-parser/lark/pull/993

Regarding 

- `__visit_tokens__` - To be honest I'm not sure why it bothered us. I think it can stay the way it is.

- deprecation of lexer callbacks - I used to think that maybe the `transformer` parameter should be enough, since we added support for tokens. But it's still useful for going over ignored tokens, so unless we let `transformer` run on ignored tokens too, I think we should leave it.

- `UnexpectedToken.expected` - Maybe it's still worth keeping? I have no strong opinion atm

**Comment by erezsh** [comment]
Btw, @chanicpanic & @MegaIng,

I think we should just merge this, and continue work over the `master` branch. Any objections?

**Comment by MegaIng** [comment]
That would mean that we are not going to release any other version before 1.0, not even a `0.12.1` release. But that would be fine with me.

**Comment by erezsh** [comment]
We can always continue working on a 0.12 branch. But I'm not sure that we need to.

**Comment by MegaIng** [comment]
Fair enough. Merge is good IMO.

**Comment by erezsh** [comment]
> Or do we also want to rename to main in the process?

Seems pointless to me

**Comment by MegaIng** [comment]
>Seems pointless to me

Fair enough. I dont have a strong opinion either, but this would be a prime opportunity.

**Comment by chanicpanic** [comment]
No objections to merging from me.

**Comment by erezsh** [comment]
Alright. I'll do it tomorrow then.

**Comment by erezsh** [comment]
Lark `master` is now officially Python 3 only!

**Pull request merged**